src/routing/executor.rs

```diff
@@ -6,101 +6,81 @@
 // permissions and limitations relating to use of the SAFE Network Software.
 
-use super::{Command, Stage};
+use super::{comm::IncomingMessages, Command, Stage};
 use crate::{event::Event, messages::Message};
 use bytes::Bytes;
-use qp2p::{IncomingConnections, IncomingMessages, Message as QuicP2pMsg};
-use std::{future::Future, net::SocketAddr, sync::Arc};
-use tokio::sync::watch;
+use std::{net::SocketAddr, sync::Arc};
+use tokio::{sync::oneshot, task};
 
 pub struct Executor {
-    cancel_tx: watch::Sender<bool>,
+    cancel_tx: Option<oneshot::Sender<()>>,
 }
 
 impl Drop for Executor {
     fn drop(&mut self) {
-        let _ = self.cancel_tx.broadcast(true);
+        if let Some(tx) = self.cancel_tx.take() {
+            let _ = tx.send(());
+        }
     }
 }
 
 impl Executor {
-    pub(crate) async fn new(stage: Arc<Stage>, incoming_conns: IncomingConnections) -> Self {
-        let (cancel_tx, mut cancel_rx) = watch::channel(false);
-
-        // Need to `recv` once, otherwise we would cancel all the tasks immediatelly
-        // (for more details, see: https://docs.rs/tokio/0.2.22/tokio/sync/watch/struct.Receiver.html?search=#method.recv).
-        let _ = cancel_rx.recv().await;
+    pub(crate) async fn new(stage: Arc<Stage>, incoming_msgs: IncomingMessages) -> Self {
+        let (cancel_tx, cancel_rx) = oneshot::channel();
 
-        spawn_connections_handler(stage, incoming_conns, cancel_rx);
+        let _ = task::spawn(async move {
+            tokio::select! {
+                _ = handle_incoming_messages(stage, incoming_msgs) => (),
+                _ = cancel_rx => (),
+            }
+        });
 
-        Self { cancel_tx }
+        Self {
+            cancel_tx: Some(cancel_tx),
+        }
     }
 }
 
-// Spawns a task which handles each new incoming connection from peers
-fn spawn_connections_handler(
-    stage: Arc<Stage>,
-    mut incoming_conns: IncomingConnections,
-    cancel_rx: watch::Receiver<bool>,
-) {
-    let _ = tokio::spawn(cancellable(cancel_rx.clone(), async move {
-        while let Some(incoming_msgs) = incoming_conns.next().await {
-            trace!(
-                "New connection established by peer {}",
-                incoming_msgs.remote_addr()
-            );
-            spawn_messages_handler(stage.clone(), incoming_msgs, cancel_rx.clone())
-        }
-    }));
-}
+async fn handle_incoming_messages(stage: Arc<Stage>, mut incoming_msgs: IncomingMessages) {
+    while let Some(msg) = incoming_msgs.next().await {
+        match msg {
+            qp2p::Message::UniStream { bytes, src, .. } => {
+                trace!(
+                    "New message ({} bytes) received on a uni-stream from: {}",
+                    bytes.len(),
+                    src
+                );
+                // Since it's arriving on a uni-stream we treat it as a Node
+                // message which we need to be processed by us, as well as
```

Comment:
treat it as a Node message which need to be processed by us, and report it to the event stream consumer as well.